Mempool `tx.Hash` nil pointer dereference in frontend/service.go:480

lightwalletd crashed a few times. Restarting seems to get it going again and it's running again for now.

This is on mainnet with the debian 5.1.0 release of zcashd and db2795aad7043fb226bbf579aee29633583990ef version of lightwalletd.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0xa1009c]

goroutine 133 [running]:
github.com/zcash/lightwalletd/frontend.(*lwdStreamer).GetMempoolTx(0xc00000f7a0, 0xc000d86500, {0xd513e0, 0xc0002688f0})
	free2z/zcash/lightwalletd/frontend/service.go:480 +0x3fc
github.com/zcash/lightwalletd/walletrpc._CompactTxStreamer_GetMempoolTx_Handler({0xb51a00, 0xc0014f1340}, {0xd4de80, 0xc00000f7a0})
	free2z/zcash/lightwalletd/walletrpc/service_grpc.pb.go:632 +0xd0
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).StreamServerInterceptor.func1({0xb51a00, 0xc0014f1340}, {0xd4dfa0, 0xc00024a5b0}, 0x8, 0xc85fb0)
	go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:121 +0x109
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1.1.1({0xb51a00, 0xc0014f1340}, {0xd4dfa0, 0xc00024a5b0})
	go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:49 +0x3a
github.com/grpc-ecosystem/go-grpc-middleware.ChainStreamServer.func1({0xb51a00, 0xc0014f1340}, {0xd4dfa0, 0xc00024a5b0}, 0xa5eae0, 0xc0002688e0)
	go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.3.0/chain.go:58 +0xbf
google.golang.org/grpc.(*Server).processStreamingRPC(0xc000238540, {0xd54568, 0xc000285a00}, 0xc000263b00, 0xc0002a3170, 0x11f2c00, 0x0)
	go/pkg/mod/google.golang.org/grpc@v1.47.0/server.go:1549 +0xe8e
google.golang.org/grpc.(*Server).handleStream(0xc000238540, {0xd54568, 0xc000285a00}, 0xc000263b00, 0x0)
	go/pkg/mod/google.golang.org/grpc@v1.47.0/server.go:1624 +0x9e5
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	go/pkg/mod/google.golang.org/grpc@v1.47.0/server.go:922 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	go/pkg/mod/google.golang.org/grpc@v1.47.0/server.go:920 +0x294
```

I haven't looked into it yet but it's the `tx.Hash` in:

```
	for _, txid := range MempoolFilter(mempoolList, excludeHex) {
		tx := (*mempoolMap)[txid]
		if len(tx.Hash) > 0 {
```

--- a/frontend/service.go
+++ b/frontend/service.go
@@ -15,6 +15,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"sync"
 	"sync/atomic"
 	"time"
 
@@ -27,6 +28,7 @@ type lwdStreamer struct {
 	cache      *common.BlockCache
 	chainName  string
 	pingEnable bool
+	mutex      sync.Mutex
 	walletrpc.UnimplementedCompactTxStreamerServer
 }
 
@@ -57,12 +59,15 @@ func checkTaddress(taddr string) error {
 
 // GetLatestBlock returns the height of the best chain, according to zcashd.
 func (s *lwdStreamer) GetLatestBlock(ctx context.Context, placeholder *walletrpc.ChainSpec) (*walletrpc.BlockID, error) {
+	// Lock to ensure we return consistent height and hash
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 	latestBlock := s.cache.GetLatestHeight()
-	latestHash := s.cache.GetLatestHash()
 
 	if latestBlock == -1 {
 		return nil, errors.New("cache is empty, server is probably not yet ready")
 	}
+	latestHash := s.cache.GetLatestHash()
 
 	return &walletrpc.BlockID{Height: uint64(latestBlock), Hash: latestHash}, nil
 }
@@ -151,11 +156,10 @@ func (s *lwdStreamer) GetBlock(ctx context.Context, id *walletrpc.BlockID) (*wal
 // 'end' inclusively.
 func (s *lwdStreamer) GetBlockRange(span *walletrpc.BlockRange, resp walletrpc.CompactTxStreamer_GetBlockRangeServer) error {
 	blockChan := make(chan *walletrpc.CompactBlock)
-	errChan := make(chan error)
 	if span.Start == nil || span.End == nil {
 		return errors.New("must specify start and end heights")
 	}
-
+	errChan := make(chan error)
 	go common.GetBlockRange(s.cache, blockChan, errChan, int(span.Start.Height), int(span.End.Height))
 
 	for {
@@ -410,6 +414,9 @@ var mempoolList []string
 var lastMempool time.Time
 
 func (s *lwdStreamer) GetMempoolTx(exclude *walletrpc.Exclude, resp walletrpc.CompactTxStreamer_GetMempoolTxServer) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
 	if time.Since(lastMempool).Seconds() >= 2 {
 		lastMempool = time.Now()
 		// Refresh our copy of the mempool.

